core/src/main/scala/org/apache/spark/status/AppStatusListener.scala

```diff
@@ -772,6 +772,11 @@ private[spark] class AppStatusListener(
     event.maxOnHeapMem.foreach { _ =>
       exec.totalOnHeap = event.maxOnHeapMem.get
       exec.totalOffHeap = event.maxOffHeapMem.get
+      // SPARK-30594: whenever(first time or re-register) a BlockManager added, all blocks
+      // from this BlockManager will be reported to driver later. So, we should clean up
+      // used memory to avoid overlapped count.
+      exec.usedOnHeap = 0
+      exec.usedOffHeap = 0
```

Comment:
Can we add a UT for this code logic?

Reply:
It should be easy to add a new test case in `AppStatusListenerSuite` to enforce the behavior.

Reply:
Done in https://github.com/apache/spark/pull/27418/commits/f3fabdc812917c3e13283d380b99656ecc06cd93.